### PR TITLE
LSIF: Defunctionize closest dump

### PR DESCRIPTION
**Pull request by efritz**
Remove the closest_dump function from the db layer and put it into lsif-server where the rest of the similar logic is. This allows us to share some additional code, and benchmarks shows that there is no meaningful performance regression.

### CODE REVIEW

File: lsif/src/shared/models/queries.ts [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
+            SELECT * FROM (
+                WITH l_inner AS (SELECT * FROM lineage)
+                -- get next ancestor
+                SELECT c.*, 'A' FROM l_inner l JOIN lsif_commits c ON l.direction = 'A' AND c.repository = l.repository AND c."commit" = l.parent_commit
+                UNION
+                -- get next descendant
+                SELECT c.*, 'D' FROM l_inner l JOIN lsif_commits c ON l.direction = 'D' and c.repository = l.repository AND c.parent_commit = l."commit"
+            ) subquery
+        )
+    `
```
Comment (keegancsmith) [inline lsif/src/shared/models/queries.ts]: image

Comment (tsenart) [reply lsif/src/shared/models/queries.ts]: xD

File: lsif/src/shared/models/queries.ts [hunk 2]
```diff
@@ -1,14 +1,58 @@
...
+            SELECT c.* FROM lineage a JOIN lsif_commits c ON a.repository = c.repository AND a.parent = c."commit"
+        )
+    `
+}
+
+/**
+ * Return a recursive CTE `lineage` that returns ancestors and descendants of the commit for
+ * the given repository. This assumes that the repo name is $1 and the commit is $2. This
+ * happens to evaluate in Postgres as a lazy generator, which allows us to pull the "next"
+ * closest commit  in either direction from the source commit as needed.
```
Comment (chrismwendt) [inline lsif/src/shared/models/queries.ts]:
```suggestion
 * closest commit in either direction from the source commit as needed.
```

File: lsif/src/shared/models/queries.ts [hunk 3]
```diff
@@ -1,14 +1,58 @@
...
+            UNION
+            SELECT c.* FROM lineage a JOIN lsif_commits c ON a.repository = c.repository AND a.parent = c."commit"
+        )
+    `
+}
+
+/**
+ * Return a recursive CTE `lineage` that returns ancestors and descendants of the commit for
+ * the given repository. This assumes that the repo name is $1 and the commit is $2. This
+ * happens to evaluate in Postgres as a lazy generator, which allows us to pull the "next"
```
Comment (chrismwendt) [inline lsif/src/shared/models/queries.ts]: I recall that we were not certain whether or not the newer version of Postgres made this efficient (you're on 12?). I'd check if this is also efficient on Postgres 9.6, our lowest supported version https://docs.sourcegraph.com/admin/postgres#version-requirements

File: lsif/src/shared/models/queries.ts [hunk 4]
```diff
@@ -1,14 +1,58 @@
...
+                UNION
+                -- seed recursive set with commit looking in descendant direction
+                SELECT c.*, 'D' FROM lsif_commits c WHERE c.repository = $1 AND c."commit" = $2
+            ) l
+
+            UNION
+
+            SELECT * FROM (
+                WITH l_inner AS (SELECT * FROM lineage)
+                -- get next ancestor
```
Comment (chrismwendt) [inline lsif/src/shared/models/queries.ts]:
```suggestion
                -- get next ancestors (multiple parents for merge commits)
```

File: lsif/src/shared/models/queries.ts [hunk 5]
```diff
@@ -1,14 +1,58 @@
...
+                SELECT c.*, 'A' FROM l_inner l JOIN lsif_commits c ON l.direction = 'A' AND c.repository = l.repository AND c."commit" = l.parent_commit
+                UNION
+                -- get next descendant
```
Comment (chrismwendt) [inline lsif/src/shared/models/queries.ts]:
```suggestion
                -- get next descendants
```

**Comment by efritz** [comment]
> Approve for the parts owned by core services (excluding the SQL).

LOL, the bindata?

File: lsif/src/shared/models/queries.ts (hunk 3 above)
Comment (efritz) [reply lsif/src/shared/models/queries.ts]: With 9.6 in a container, limit=2500.

Screen Shot 2020-01-08 at 10 38 36 AM

**Pull request merged**